GitHub login broken multiple ways

While The GitHub login authorized callback returns JSON. It needs to return a redirect.

After fixing that, I noticed that auth-github-login is being called twice:

```txt
10.10.10.1 - - [11/Feb/2020 03:04:14] "GET /auth-github-login/elegant HTTP/1.1" 302 -
10.10.10.1 - - [11/Feb/2020 03:04:14] "GET /auth-github-login/elegant HTTP/1.1" 302 -
```

The first and second call locations are [the same location](https://github.com/radiasoft/sirepo/blob/master/sirepo/package_data/static/js/sirepo.js#L2879):

```txt
VM2346 angular.min.js:127 Error: hello
    at Object.self.globalRedirect (VM2352 sirepo.js:1648)
    at new <anonymous> (VM2352 sirepo.js:2877)
    at Object.instantiate (VM2346 angular.min.js:45)
    at VM2346 angular.min.js:98
    at Object.link (VM2362 angular-route.min.js:8)
    at VM2346 angular.min.js:17
    at Ba (VM2346 angular.min.js:89)
    at p (VM2346 angular.min.js:73)
    at g (VM2346 angular.min.js:64)
    at VM2346 angular.min.js:64 "<div data-ng-view="" class="sr-view-content">"
```

The LoginWithController seems to be called twice or somethings wrong with doing a redirect in the middle of a controller initialization. Definitely something angular I don't understand.

--- a/etc/run-auth-email.sh
+++ b/etc/run-auth-email.sh
@@ -5,6 +5,6 @@ export SIREPO_AUTH_EMAIL_SMTP_PASSWORD='n/a'
 # POSIT: same as sirepo.auth.email._DEV_SMTP_SERVER
 export SIREPO_AUTH_EMAIL_SMTP_SERVER='dev'
 export SIREPO_AUTH_EMAIL_SMTP_USER='n/a'
-export SIREPO_AUTH_METHODS='email:guest'
+export SIREPO_AUTH_METHODS='email:guest:github'
 export PYKERN_PKDEBUG_WANT_PID_TIME=1
 exec sirepo service http

--- a/setup.py
+++ b/setup.py
@@ -13,7 +13,7 @@ pykern.pksetup.setup(
     install_requires=[
         # some "concrete" dependencies in requirements.txt
         'Flask>=1.1',
-        'Flask-OAuthlib; python_version < "3"',
+#        'Flask-OAuthlib; python_version < "3"',
         'Flask-Mail; python_version < "3"',
         'Flask_SQLAlchemy',
         'SQLAlchemy; python_version < "3"',

--- a/setup.py
+++ b/setup.py
@@ -13,10 +13,10 @@ pykern.pksetup.setup(
     install_requires=[
         # some "concrete" dependencies in requirements.txt
         'Flask>=1.1',
-#        'Flask-OAuthlib; python_version < "3"',
         'Flask-Mail; python_version < "3"',
         'Flask_SQLAlchemy',
         'SQLAlchemy; python_version < "3"',
+        'Authlib>=0.13',
         'aenum',
         'beaker; python_version < "3"',
         'celery==3.1.23; python_version < "3"',
@@ -31,10 +31,6 @@ pykern.pksetup.setup(
         'pyIsEmail',
         'pykern',
         'pytz==2015.7',
-        # requests-oauthlib 1.3.0 has requirement oauthlib>=3.0.0, but you'll have oauthlib 2.1.0 which is incompatible
-        'requests-oauthlib==1.1.0; python_version < "3"',
-        # for Flask-OAuthlib
-        'oauthlib!=2.0.3,!=2.0.4,!=2.0.5,<3.0.0,>=1.1.2; python_version < "3"',
         'scikit-learn==0.20',
         'sympy; python_version < "3"',
         'user-agents',

--- a/sirepo/auth/__init__.py
+++ b/sirepo/auth/__init__.py
@@ -16,6 +16,7 @@ from sirepo import http_reply
 from sirepo import http_request
 from sirepo import auth_db
 from sirepo import util
+import sirepo.uri
 import sirepo.template
 import datetime
 import importlib
@@ -175,7 +176,7 @@ def logged_in_user():
     return res
 
 
-def login(module, uid=None, model=None, sim_type=None, display_name=None, is_mock=False):
+def login(module, uid=None, model=None, sim_type=None, display_name=None, is_mock=False, want_redirect=False):
     """Login the user
 
     Raises an exception if successful, except in the case of methods
@@ -229,7 +230,7 @@ def login(module, uid=None, model=None, sim_type=None, display_name=None, is_moc
     if sim_type:
         if guest_uid and guest_uid != uid:
             simulation_db.move_user_simulations(guest_uid, uid)
-        login_success_response(sim_type)
+        login_success_response(sim_type, want_redirect)
     assert not module.AUTH_METHOD_VISIBLE
 
 
@@ -248,16 +249,19 @@ def login_fail_redirect(sim_type=None, module=None, reason=None, reload_js=False
     )
 
 
-def login_success_response(sim_type):
+def login_success_response(sim_type, want_redirect=False):
     r = None
-    if cookie.get_value(_COOKIE_STATE) == _STATE_COMPLETE_REGISTRATION:
-        if cookie.get_value(_COOKIE_METHOD) == METHOD_GUEST:
-            complete_registration()
-        else:
-            r = 'completeRegistration'
+    if (
+        cookie.get_value(_COOKIE_STATE) == _STATE_COMPLETE_REGISTRATION
+        and cookie.get_value(_COOKIE_METHOD) == METHOD_GUEST
+    ):
+        complete_registration()
+    if want_redirect:
+        raise sirepo.util.Redirect(sirepo.uri.app_root(sim_type))
     raise sirepo.util.Response(
         response=http_reply.gen_json_ok(PKDict(authState=_auth_state())),
     )
+
 
 def need_complete_registration(model):
     """Does unauthenticated user need to complete registration?

--- a/sirepo/auth/github.py
+++ b/sirepo/auth/github.py
@@ -147,7 +147,7 @@ def _init_model(db, base):
 
 
 def _oauth_client():
-    return authlib.flask.client.OAuth(flask.current_app).remote_app(
+    return authlib.integrations.flask_client.OAuth(flask.current_app).remote_app(
         'github',
         consumer_key=cfg.key,
         consumer_secret=cfg.secret,

--- a/sirepo/auth/github.py
+++ b/sirepo/auth/github.py
@@ -77,7 +77,7 @@ def api_authGithubAuthorized():
         else:
             u = AuthGithubUser(oauth_id=d['id'], user_name=d['login'])
         u.save()
-        auth.login(this_module, model=u, sim_type=t)
+        auth.login(this_module, model=u, sim_type=t, want_redirect=True)
         raise AssertionError('auth.login returned unexpectedly')
 
 
@@ -153,8 +153,7 @@ def _init_model(db, base):
 
 
 def _oauth_client():
-    authlib.integrations.requests_client.OAuth2Session
-    r = authlib.integrations.flask_client.OAuth(app, cache=_OAuthCache)
+    r = authlib.integrations.flask_client.OAuth(_app)
     r.register(
         access_token_params=None,
         access_token_url='https://github.com/login/oauth/access_token',

--- a/sirepo/auth/github.py
+++ b/sirepo/auth/github.py
@@ -20,7 +20,7 @@ from sirepo import auth_db
 from sirepo import util
 import flask
 import flask.sessions
-import flask_oauthlib.client
+import authlib.flask.client
 import sirepo.template
 
 AUTH_METHOD = 'github'
@@ -147,7 +147,7 @@ def _init_model(db, base):
 
 
 def _oauth_client():
-    return flask_oauthlib.client.OAuth(flask.current_app).remote_app(
+    return authlib.flask.client.OAuth(flask.current_app).remote_app(
         'github',
         consumer_key=cfg.key,
         consumer_secret=cfg.secret,

--- a/sirepo/auth/github.py
+++ b/sirepo/auth/github.py
@@ -20,8 +20,9 @@ from sirepo import auth_db
 from sirepo import util
 import flask
 import flask.sessions
-import authlib.flask.client
+import authlib.integrations.flask_client
 import sirepo.template
+
 
 AUTH_METHOD = 'github'
 
@@ -37,9 +38,13 @@ UserModel = None
 #: module handle
 this_module = pkinspect.this_module()
 
-# cookie keys for github (prefix is "srag")
+#: cookie keys for github (prefix is "srag")
 _COOKIE_NONCE = 'sragn'
 _COOKIE_SIM_TYPE = 'srags'
+
+
+#: cached for _oauth_client
+_app = None
 
 
 @api_perm.allow_cookieless_set_user
@@ -52,8 +57,7 @@ def api_authGithubAuthorized():
     expect = cookie.unchecked_remove(_COOKIE_NONCE) or '<missing-nonce>'
     t = cookie.unchecked_remove(_COOKIE_SIM_TYPE)
     oc = _oauth_client()
-    resp = oc.authorized_response()
-    if not resp:
+    if not oc.authorize_access_token():
         util.raise_forbidden('missing oauth response')
     got = flask.request.args.get('state', '<missing-state>')
     if expect != got:
@@ -64,7 +68,7 @@ def api_authGithubAuthorized():
         )
         auth.login_fail_redirect(t, this_module, 'oauth-state', reload_js=True)
         raise AssertionError('auth.login_fail_redirect returned unexpectedly')
-    d = oc.get('user', token=(resp['access_token'], '')).data
+    d = oc.get('user').json()
     with auth_db.thread_lock:
         u = AuthGithubUser.search_by(oauth_id=d['id'])
         if u:
@@ -88,7 +92,7 @@ def api_authGithubLogin(simulation_type):
         # must be executed in an app and request context so can't
         # initialize earlier.
         cfg.callback_uri = uri_router.uri_for_api('authGithubAuthorized')
-    return _oauth_client().authorize(callback=cfg.callback_uri, state=s)
+    return _oauth_client().authorize_redirect(redirect_uri=cfg.callback_uri, state=s)
 
 
 @api_perm.allow_cookieless_set_user
@@ -105,14 +109,15 @@ def avatar_uri(model, size):
 
 
 def init_apis(app, *args, **kwargs):
-    global cfg
+    global cfg, _app
     cfg = pkconfig.init(
         key=pkconfig.Required(str, 'Github key'),
         secret=pkconfig.Required(str, 'Github secret'),
         callback_uri=(None, str, 'Github callback URI (defaults to api_authGithubAuthorized)'),
     )
-    app.session_interface = _FlaskSessionInterface()
     auth_db.init_model(app, _init_model)
+    app.session_interface = _FlaskSessionInterface()
+    _app = app
 
 
 class _FlaskSession(dict, flask.sessions.SessionMixin):
@@ -120,6 +125,7 @@ class _FlaskSession(dict, flask.sessions.SessionMixin):
 
 
 class _FlaskSessionInterface(flask.sessions.SessionInterface):
+
     """Emphemeral session for oauthlib.client state
 
     Without this class, Flask creates a NullSession which can't
@@ -147,13 +153,17 @@ def _init_model(db, base):
 
 
 def _oauth_client():
-    return authlib.integrations.flask_client.OAuth(flask.current_app).remote_app(
-        'github',
-        consumer_key=cfg.key,
-        consumer_secret=cfg.secret,
-        base_url='https://api.github.com/',
-        request_token_url=None,
-        access_token_method='POST',
+    authlib.integrations.requests_client.OAuth2Session
+    r = authlib.integrations.flask_client.OAuth(app, cache=_OAuthCache)
+    r.register(
+        access_token_params=None,
         access_token_url='https://github.com/login/oauth/access_token',
+        api_base_url='https://api.github.com/',
+        authorize_params=None,
         authorize_url='https://github.com/login/oauth/authorize',
+        client_id=cfg.key,
+        client_kwargs={'scope': 'user:email'},
+        client_secret=cfg.secret,
+        name='github',
     )
+    return r.create_client('github')

--- a/sirepo/github_srunit.py
+++ b/sirepo/github_srunit.py
@@ -14,7 +14,7 @@ class MockOAuthClient(object):
 
         self.values = pkcollections.Dict(
             access_token='xyzzy',
-            data=pkcollections.Dict(
+            user=pkcollections.Dict(
                 # don't really care about id as long as it is bound to login
                 id=user_name,
                 login=user_name,
@@ -25,7 +25,7 @@ class MockOAuthClient(object):
     def __call__(self, *args, **kwargs):
         return self
 
-    def authorize(self, callback, state):
+    def authorize_redirect(self, redirect_uri, state):
         from sirepo.auth import github
         import sirepo.http_reply
 
@@ -39,8 +39,14 @@ class MockOAuthClient(object):
             ),
         )
 
-    def authorized_response(self, *args, **kwargs):
+    def authorize_access_token(self, *args, **kwargs):
         return self.values
 
     def get(self, *args, **kwargs):
-        return self.values
+        return _JSON(self.values[args[0]])
+
+
+class _JSON(PKDict):
+
+    def json(self):
+        return self

--- a/sirepo/github_srunit.py
+++ b/sirepo/github_srunit.py
@@ -5,6 +5,8 @@ u"""support for oauth test
 :license: http://www.apache.org/licenses/LICENSE-2.0.html
 """
 from __future__ import absolute_import, division, print_function
+from pykern.pkcollections import PKDict
+
 
 class MockOAuthClient(object):
 
@@ -29,7 +31,7 @@ class MockOAuthClient(object):
         from sirepo.auth import github
         import sirepo.http_reply
 
-        self.values.callback = callback
+        self.values.redirect_uri = redirect_uri
         self.values.state = state
         return sirepo.http_reply.gen_redirect(
             'https://github.com/login/oauth/oauthorize?response_type=code&client_id={}&redirect_uri={}&state={}'.format(

--- a/sirepo/package_data/static/js/sirepo.js
+++ b/sirepo/package_data/static/js/sirepo.js
@@ -183,7 +183,7 @@ SIREPO.app.factory('activeSection', function(authState, requestSender, $location
     return self;
 });
 
-SIREPO.app.factory('appState', function(errorService, fileManager, requestQueue, requestSender, $document, $interval, $rootScope, $window) {
+SIREPO.app.factory('appState', function(errorService, fileManager, requestQueue, requestSender, $document, $interval, $rootScope) {
     var self = {
         models: {},
     };
@@ -1468,7 +1468,7 @@ SIREPO.app.factory('panelState', function(appState, requestSender, simulationQue
     return self;
 });
 
-SIREPO.app.factory('requestSender', function(cookieService, errorService, localRoutes, $http, $location, $interval, $q, $rootScope, $window, $route) {
+SIREPO.app.factory('requestSender', function(cookieService, errorService, localRoutes, $http, $location, $interval, $q, $rootScope, $window) {
     var self = {};
     var HTML_TITLE_RE = new RegExp('>([^<]+)</', 'i');
     var IS_HTML_ERROR_RE = new RegExp('^(?:<html|<!doctype)', 'i');
@@ -1644,6 +1644,11 @@ SIREPO.app.factory('requestSender', function(cookieService, errorService, localR
             $location.hash('#');
         }
         $window.location.href = u;
+        // The whole app will reload.
+        // Don't respond to additional location change events from the current app.
+        $rootScope.$on('$locationChangeStart', function (event) {
+            event.preventDefault();
+        });
     };
 
     self.globalRedirectRoot = function() {
@@ -2845,7 +2850,7 @@ SIREPO.app.controller('LoginController', function (authService, authState, reque
     }
 });
 
-SIREPO.app.controller('LoginWithController', function ($route, $window, errorService, appState, authState, requestSender) {
+SIREPO.app.controller('LoginWithController', function (authState, errorService, requestSender, $route) {
     var self = this;
     var m = $route.current.params.method || '';
     self.showWarning = false;
@@ -2880,7 +2885,7 @@ SIREPO.app.controller('LoginWithController', function ($route, $window, errorSer
     }
 });
 
-SIREPO.app.controller('LoginConfirmController', function ($route, $window, authState, requestSender) {
+SIREPO.app.controller('LoginConfirmController', function (authState, requestSender, $route) {
     var self = this;
     var p = $route.current.params;
     self.data = {};
@@ -2986,7 +2991,7 @@ SIREPO.app.controller('ServerUpgradedController', function (errorService, reques
     requestSender.globalRedirectRoot();
 });
 
-SIREPO.app.controller('SimulationsController', function (appState, cookieService, errorService, fileManager, notificationService, panelState, requestSender, $location, $rootScope, $sce, $scope, $window) {
+SIREPO.app.controller('SimulationsController', function (appState, cookieService, errorService, fileManager, notificationService, panelState, requestSender, $location, $rootScope, $sce, $scope) {
     var self = this;
 
     $rootScope.$broadcast('simulationUnloaded');

